### ISSUE TITLE
Unable to shutdown Tomcat in 3.4

### ISSUE THREAD

**Issue by PRESENTERS** [gitbucket/gitbucket#798]
I have deployed GitBucket to Tomcat, but the following error occurred when I shut down the Tomcat.
When this error occurs, Tomcat's process will continuously stay in the OS.

CentOS 7.0
Tomcat 7.0.62
OpenJDK 1.8
GitBucket 3.4

```
Jun 28, 2015 11:02:23 PM org.apache.catalina.core.StandardServer await
INFO: A valid shutdown command was received via the shutdown port. Stopping the Server instance.
Jun 28, 2015 11:02:23 PM org.apache.coyote.AbstractProtocol pause
INFO: Pausing ProtocolHandler ["http-bio-8180"]
Jun 28, 2015 11:02:23 PM org.apache.coyote.AbstractProtocol pause
INFO: Pausing ProtocolHandler ["ajp-bio-8009"]
Jun 28, 2015 11:02:23 PM org.apache.catalina.core.StandardService stopInternal
INFO: Stopping service Catalina
Jun 28, 2015 11:02:23 PM grizzled.slf4j.Logger info
INFO: Destroying life cycle class: ScalatraBootstrap
Jun 28, 2015 11:02:23 PM org.apache.catalina.loader.WebappClassLoader clearReferencesJdbc
SEVERE: The web application [/gitbucket] registered the JDBC driver [org.h2.Driver] but failed to unregister it when the web application was stopped. To prevent a memory leak, the JDBC Driver has been forcibly unregistered.
Jun 28, 2015 11:02:23 PM org.apache.catalina.loader.WebappClassLoader clearReferencesThreads
SEVERE: The web application [/gitbucket] appears to have started a thread named [Timer-0] but has failed to stop it. This is very likely to create a memory leak.
Jun 28, 2015 11:02:23 PM org.apache.catalina.loader.WebappClassLoader clearReferencesThreads
SEVERE: The web application [/gitbucket] appears to have started a thread named [MVStore background writer nio:/opt/apache-tomcat-7.0.62/.gitbucket/data.mv.db] but has failed to stop it. This is very likely to create a memory leak.
Jun 28, 2015 11:02:23 PM org.apache.catalina.loader.WebappClassLoader clearReferencesThreads
SEVERE: The web application [/gitbucket] appears to have started a thread named [job-scheduler-1] but has failed to stop it. This is very likely to create a memory leak.
Jun 28, 2015 11:02:23 PM org.apache.catalina.loader.WebappClassLoader clearReferencesThreads
SEVERE: The web application [/gitbucket] appears to have started a thread named [job-akka.actor.default-dispatcher-5] but has failed to stop it. This is very likely to create a memory leak.
Jun 28, 2015 11:02:23 PM org.apache.catalina.loader.WebappClassLoader clearReferencesThreads
SEVERE: The web application [/gitbucket] appears to have started a thread named [AKKA_QRTZ_-1] but has failed to stop it. This is very likely to create a memory leak.
Jun 28, 2015 11:02:23 PM org.apache.catalina.loader.WebappClassLoader clearReferencesThreads
SEVERE: The web application [/gitbucket] appears to have started a thread named [Timer-1] but has failed to stop it. This is very likely to create a memory leak.
Jun 28, 2015 11:02:23 PM org.apache.catalina.loader.WebappClassLoader clearReferencesThreads
SEVERE: The web application [/gitbucket] appears to have started a thread named [Resource Destroyer in BasicResourcePool.close()] but has failed to stop it. This is very likely to create a memory leak.
Jun 28, 2015 11:02:23 PM org.apache.catalina.loader.WebappClassLoader clearReferencesThreads
SEVERE: The web application [/gitbucket] appears to have started a thread named [QuartzScheduler~job_QuartzSchedulerThread] but has failed to stop it. This is very likely to create a memory leak.
Jun 28, 2015 11:02:23 PM org.apache.coyote.AbstractProtocol stop
INFO: Stopping ProtocolHandler ["http-bio-8180"]
Jun 28, 2015 11:02:23 PM org.apache.coyote.AbstractProtocol stop
INFO: Stopping ProtocolHandler ["ajp-bio-8009"]
Jun 28, 2015 11:02:23 PM org.apache.coyote.AbstractProtocol destroy
INFO: Destroying ProtocolHandler ["http-bio-8180"]
Jun 28, 2015 11:02:23 PM org.apache.coyote.AbstractProtocol destroy
INFO: Destroying ProtocolHandler ["ajp-bio-8009"]
Jun 28, 2015 11:02:23 PM org.apache.catalina.loader.WebappClassLoader loadClass
INFO: Illegal access: this web application instance has been stopped already.  Could not load com.mchange.v2.resourcepool.BasicResourcePool$1DestroyResourceTask.  The eventual following stack trace is caused by an error thrown for debugging purposes as well as to attempt to terminate the thread which caused the illegal access, and has no functional impact.
java.lang.IllegalStateException
        at org.apache.catalina.loader.WebappClassLoader.loadClass(WebappClassLoader.java:1612)
        at org.apache.catalina.loader.WebappClassLoader.loadClass(WebappClassLoader.java:1571)
        at com.mchange.v2.resourcepool.BasicResourcePool.destroyResource(BasicResourcePool.java:980)
        at com.mchange.v2.resourcepool.BasicResourcePool.access$100(BasicResourcePool.java:32)
        at com.mchange.v2.resourcepool.BasicResourcePool$5.run(BasicResourcePool.java:1174)

Exception in thread "Resource Destroyer in BasicResourcePool.close()" java.lang.NoClassDefFoundError: com/mchange/v2/resourcepool/BasicResourcePool$1DestroyResourceTask
        at com.mchange.v2.resourcepool.BasicResourcePool.destroyResource(BasicResourcePool.java:980)
        at com.mchange.v2.resourcepool.BasicResourcePool.access$100(BasicResourcePool.java:32)
        at com.mchange.v2.resourcepool.BasicResourcePool$5.run(BasicResourcePool.java:1174)
Caused by: java.lang.ClassNotFoundException: com.mchange.v2.resourcepool.BasicResourcePool$1DestroyResourceTask
        at org.apache.catalina.loader.WebappClassLoader.loadClass(WebappClassLoader.java:1720)
        at org.apache.catalina.loader.WebappClassLoader.loadClass(WebappClassLoader.java:1571)
        ... 3 more
```

**Comment by he64-ktgw** [comment]
I'm having the same problem.

Windows 7 64bit
XAMMP 5.6.15
Tomcat 7.0.56
JDK 1.8.0 66
GitBucket 3.9

12 21, 2015 10:41:42 午後 org.apache.catalina.core.StandardServer await
情報: A valid shutdown command was received via the shutdown port. Stopping the
Server instance.
12 21, 2015 10:41:42 午後 org.apache.coyote.AbstractProtocol pause
情報: Pausing ProtocolHandler ["http-bio-8080"]
12 21, 2015 10:41:42 午後 org.apache.coyote.AbstractProtocol pause
情報: Pausing ProtocolHandler ["ajp-bio-8009"]
12 21, 2015 10:41:42 午後 org.apache.catalina.core.StandardService stopInternal
情報: サービス Catalina を停止します
12 21, 2015 10:41:42 午後 grizzled.slf4j.Logger info
情報: Destroying life cycle class: ScalatraBootstrap
12 21, 2015 10:41:42 午後 org.apache.catalina.loader.WebappClassLoader clearRefe
rencesJdbc
重大: The web application [/gitbucket] registered the JDBC driver [org.h2.Driver
] but failed to unregister it when the web application was stopped. To prevent a
 memory leak, the JDBC Driver has been forcibly unregistered.
12 21, 2015 10:41:42 午後 org.apache.catalina.loader.WebappClassLoader clearRefe
rencesThreads
重大: The web application [/gitbucket] appears to have started a thread named [M
VStore background writer nio:D:/devenv/.gitbucket/data.mv.db] but has failed to
stop it. This is very likely to create a memory leak.
12 21, 2015 10:41:42 午後 org.apache.catalina.loader.WebappClassLoader clearRefe
rencesThreads
重大: The web application [/gitbucket] appears to have started a thread named [j
ob-scheduler-1] but has failed to stop it. This is very likely to create a memor
y leak.
12 21, 2015 10:41:42 午後 org.apache.catalina.loader.WebappClassLoader clearRefe
rencesThreads
重大: The web application [/gitbucket] appears to have started a thread named [j
ob-akka.actor.default-dispatcher-4] but has failed to stop it. This is very like
ly to create a memory leak.
12 21, 2015 10:41:42 午後 org.apache.catalina.loader.WebappClassLoader clearRefe
rencesThreads
重大: The web application [/gitbucket] appears to have started a thread named [A
KKA_QRTZ_-1] but has failed to stop it. This is very likely to create a memory l
eak.
12 21, 2015 10:41:42 午後 org.apache.catalina.loader.WebappClassLoader clearRefe
rencesThreads
重大: The web application [/gitbucket] appears to have started a thread named [T
imer-0] but has failed to stop it. This is very likely to create a memory leak.
12 21, 2015 10:41:42 午後 org.apache.catalina.loader.WebappClassLoader clearRefe
rencesThreads
重大: The web application [/gitbucket] appears to have started a thread named [R
esource Destroyer in BasicResourcePool.close()] but has failed to stop it. This
is very likely to create a memory leak.
12 21, 2015 10:41:42 午後 org.apache.catalina.loader.WebappClassLoader clearRefe
rencesThreads
重大: The web application [/gitbucket] appears to have started a thread named [Q
uartzScheduler~job_QuartzSchedulerThread] but has failed to stop it. This is ver
y likely to create a memory leak.
12 21, 2015 10:41:42 午後 org.apache.catalina.loader.WebappClassLoader checkThre
adLocalMapForLeaks
重大: The web application [/gitbucket] created a ThreadLocal with key of type [a
kka.actor.ActorCell$$anon$1] (value [akka.actor.ActorCell$$anon$1@1b2d42e7]) and
 a value of type [scala.collection.immutable.Nil$] (value [List()]) but failed t
o remove it when the web application was stopped. Threads are going to be renewe
d over time to try and avoid a probable memory leak.
12 21, 2015 10:41:42 午後 org.apache.coyote.AbstractProtocol stop
情報: Stopping ProtocolHandler ["http-bio-8080"]
12 21, 2015 10:41:42 午後 org.apache.coyote.AbstractProtocol stop
情報: Stopping ProtocolHandler ["ajp-bio-8009"]
12 21, 2015 10:41:42 午後 org.apache.coyote.AbstractProtocol destroy
情報: Destroying ProtocolHandler ["http-bio-8080"]
12 21, 2015 10:41:42 午後 org.apache.coyote.AbstractProtocol destroy
情報: Destroying ProtocolHandler ["ajp-bio-8009"]
12 21, 2015 10:41:42 午後 org.apache.catalina.loader.WebappClassLoader loadClass

情報: 不正なアクセス: このWebアプリケーションのインスタンスは既に停止されていま
す  Could not load java.nio.channels.FileLock. 不正なアクセスを引き起こしたスレ
ッドを終了させ、投げられたエラーによりデバッグ用に次のスタックトレースが生成され
ましたが，機能に影響はありません
java.lang.IllegalStateException
        at org.apache.catalina.loader.WebappClassLoader.loadClass(WebappClassLoa
der.java:1612)
        at org.apache.catalina.loader.WebappClassLoader.loadClass(WebappClassLoa
der.java:1571)
        at org.h2.mvstore.FileStore.close(FileStore.java:185)
        at org.h2.mvstore.MVStore.closeStore(MVStore.java:788)
        at org.h2.mvstore.MVStore.close(MVStore.java:745)
        at org.h2.mvstore.db.MVTableEngine$Store.close(MVTableEngine.java:348)
        at org.h2.engine.Database.closeOpenFilesAndUnlock(Database.java:1342)
        at org.h2.engine.Database.close(Database.java:1263)
        at org.h2.engine.Database.removeSession(Database.java:1134)
        at org.h2.engine.Session.close(Session.java:676)
        at org.h2.jdbc.JdbcConnection.close(JdbcConnection.java:384)
        at com.mchange.v2.c3p0.impl.NewPooledConnection.close(NewPooledConnectio
n.java:642)
        at com.mchange.v2.c3p0.impl.NewPooledConnection.closeMaybeCheckedOut(New
PooledConnection.java:255)
        at com.mchange.v2.c3p0.impl.C3P0PooledConnectionPool$1PooledConnectionRe
sourcePoolManager.destroyResource(C3P0PooledConnectionPool.java:621)
        at com.mchange.v2.resourcepool.BasicResourcePool$1DestroyResourceTask.ru
n(BasicResourcePool.java:1065)
        at com.mchange.v2.resourcepool.BasicResourcePool.destroyResource(BasicRe
sourcePool.java:1090)
        at com.mchange.v2.resourcepool.BasicResourcePool.destroyResource(BasicRe
sourcePool.java:1051)
        at com.mchange.v2.resourcepool.BasicResourcePool.access$100(BasicResourc
ePool.java:44)
        at com.mchange.v2.resourcepool.BasicResourcePool$5.run(BasicResourcePool
.java:1305)

**Comment by swaldman** [comment]
( I've submitted a pull-request that should address c3p0-related issues with shutdown/hot-redeploy under tomcat. https://github.com/gitbucket/gitbucket/pull/1017 )

( thanks! )

**Comment by nazoking** [comment]
closed by #1017 .
